File: sdk/go/webhook.go
package iapstack

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"io"
	"mime"
	"net/http"
	"strconv"
	"strings"
	"sync"
	"time"
)

const (
	// signatureVersion identifies the authenticated webhook signing contract.
	signatureVersion = "v2"
	// signaturePrefix is the only accepted IAPStack-Signature prefix.
	signaturePrefix = signatureVersion + "="
	// signatureMACSeparator unambiguously delimits versioned MAC fields.
	signatureMACSeparator = "\n"
	// signatureBytes is the HMAC-SHA-256 digest length.
	signatureBytes = sha256.Size
	// minimumWebhookSecretBytes prevents trivially brute-forced HMAC keys.
	minimumWebhookSecretBytes = 32
	// maximumWebhookIdentityLength bounds event-controlled header identities.
	maximumWebhookIdentityLength = 128
	// defaultWebhookBodyLimit bounds one webhook payload to one mebibyte.
	defaultWebhookBodyLimit int64 = 1 << 20
	// defaultWebhookTimestampTolerance is the production five-minute replay window.
	defaultWebhookTimestampTolerance = 5 * time.Minute

	// headerEventID is the authenticated webhook event identity header.
	headerEventID = "IAPStack-Event-ID"
	// headerTimestamp is the unix-seconds webhook timestamp header.
	headerTimestamp = "IAPStack-Timestamp"
	// headerSignature is the versioned HMAC header.
	headerSignature = "IAPStack-Signature"
)

// EventStore atomically deduplicates authenticated webhook event identities.
type EventStore interface {
	// Remember records one authenticated event ID and opaque body fingerprint.
	// The verifier hashes the raw body; hosts persist the fingerprint as-is and
	// must not import crypto/sha256 to recompute it. It returns duplicate=true
	// when the same identity was stored before. A conflicting fingerprint for
	// an existing ID returns a WebhookError.
	Remember(ctx context.Context, eventID string, fingerprint []byte) (duplicate bool, err error)
}

// WebhookConfig controls signature verification, replay window, and dedupe storage.
type WebhookConfig struct {
	// Secret is the webhook signing secret; it must contain at least 32 bytes.
	Secret []byte
	// BodyLimit is the maximum accepted raw webhook body size.
	BodyLimit int64
	// TimestampTolerance is the allowed clock skew around IAPStack-Timestamp.
	TimestampTolerance time.Duration
	// Store atomically deduplicates authenticated event IDs.
	Store EventStore
}

// WebhookVerifier authenticates v2 IAPStack deliveries and deduplicates event IDs.
type WebhookVerifier struct {
	// secret is the copied HMAC key used for constant-time signature comparison.
	secret []byte
	// bodyLimit is the maximum accepted raw webhook body size.
	bodyLimit int64
	// timestampTolerance is the allowed clock skew around IAPStack-Timestamp.
	timestampTolerance time.Duration
	// store atomically deduplicates authenticated event IDs.
	store EventStore
	// clock supplies the current time for replay-window checks.
	clock func() time.Time
}

// WebhookEvent is one authenticated, optionally duplicate, entitlement change.
type WebhookEvent struct {
	// ID is the authenticated IAPStack-Event-ID value.
	ID string
	// Timestamp is the authenticated IAPStack-Timestamp value.
	Timestamp time.Time
	// Duplicate reports whether this event ID was already stored with the same body.
	Duplicate bool
	// Change is the parsed entitlement.changed payload.
	Change EntitlementChange
}

// MemoryEventStore is a process-local EventStore for tests and single-instance hosts.
type MemoryEventStore struct {
	// mu serializes identity inserts and conflict checks.
	mu sync.Mutex
	// events maps authenticated event IDs to the accepted body fingerprint.
	events map[string]string
}

// webhookHandler maps VerifyRequest onto IAPStack's outbound retry policy.
type webhookHandler struct {
	// verifier authenticates and deduplicates each delivery.
	verifier *WebhookVerifier
	// onEvent handles authenticated, non-duplicate entitlement changes.
	onEvent func(context.Context, WebhookEvent) error
}

// NewMemoryEventStore constructs an in-memory dedupe store.
func NewMemoryEventStore() *MemoryEventStore {
	return &MemoryEventStore{events: make(map[string]string)}
}

// NewWebhookVerifier validates dependencies and constructs a fail-closed verifier.
func NewWebhookVerifier(config WebhookConfig) (*WebhookVerifier, error) {
	if len(config.Secret) < minimumWebhookSecretBytes {
		return nil, errors.New("webhook signing secret must contain at least 32 bytes")
	}
	if config.Store == nil {
		return nil, errors.New("webhook event store is required")
	}
	if config.BodyLimit == 0 {
		config.BodyLimit = defaultWebhookBodyLimit
	}
	if config.BodyLimit <= 0 {
		return nil, errors.New("webhook body limit must be positive")
	}
	if config.TimestampTolerance == 0 {
		config.TimestampTolerance = defaultWebhookTimestampTolerance
	}
	if config.TimestampTolerance <= 0 {
		return nil, errors.New("webhook timestamp tolerance must be positive")
	}
	return &WebhookVerifier{
		secret:             append([]byte(nil), config.Secret...),
		bodyLimit:          config.BodyLimit,
		timestampTolerance: config.TimestampTolerance,
		store:              config.Store,
		clock:              time.Now,
	}, nil
}

// Handler returns a long-lived HTTP handler for IAPStack webhook deliveries.
//
// Construct the verifier once at process start. The handler maps signature and
// timestamp failures to 401, identity conflict to 409, and store or callback
// errors to 503. IAPStack retries only 408, 429, and 5xx; 401 is a permanent
// webhook_rejected failure.
func (verifier *WebhookVerifier) Handler(onEvent func(context.Context, WebhookEvent) error) http.Handler {
	if onEvent == nil {
		onEvent = func(context.Context, WebhookEvent) error { return nil }
	}
	return &webhookHandler{verifier: verifier, onEvent: onEvent}
}

// ServeHTTP authenticates one delivery and invokes onEvent for new events.
func (handler *webhookHandler) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
	event, err := handler.verifier.VerifyRequest(request)
	if err != nil {
		writeWebhookError(writer, err)
		return
	}
	if event.Duplicate {
		writer.WriteHeader(http.StatusNoContent)
		return
	}
	if err := handler.onEvent(request.Context(), event); err != nil {
		writeWebhookError(writer, &WebhookError{
			Code:       "receiver_unavailable",
			StatusCode: http.StatusServiceUnavailable,
			Cause:      err,
		})
		return
	}
	writer.WriteHeader(http.StatusNoContent)
}

// VerifyRequest authenticates one HTTP delivery, parses the payload, and deduplicates it.
//
// Prefer Handler so hosts do not reconstruct HTTP status mapping. Returned
// WebhookError values include the status IAPStack's outbound delivery expects.
func (verifier *WebhookVerifier) VerifyRequest(request *http.Request) (WebhookEvent, error) {
	if request == nil {
		return WebhookEvent{}, webhookFailure("invalid_body")
	}
	if request.Method != http.MethodPost {
		return WebhookEvent{}, webhookFailure("method_not_allowed")
	}
	if !isJSONContentType(request.Header.Get("Content-Type")) {
		return WebhookEvent{}, webhookFailure("content_type_required")
	}
	eventID := request.Header.Get(headerEventID)
	if !validWebhookIdentity(eventID) {
		return WebhookEvent{}, webhookFailure("invalid_event_id")
	}
	timestamp, timestampText, ok := verifier.validTimestamp(request.Header.Get(headerTimestamp))
	if !ok {
		return WebhookEvent{}, webhookFailure("invalid_timestamp")
	}
	if request.Body == nil {
		return WebhookEvent{}, webhookFailure("invalid_body")
	}
	body, err := io.ReadAll(http.MaxBytesReader(nil, request.Body, verifier.bodyLimit))
	if err != nil {
		var tooLarge *http.MaxBytesError
		if errors.As(err, &tooLarge) {
			return WebhookEvent{}, webhookFailure("body_too_large")
		}
		return WebhookEvent{}, webhookFailure("invalid_body")
	}
	if !verifier.validSignature(eventID, timestampText, body, request.Header.Get(headerSignature)) {
		return WebhookEvent{}, webhookFailure("invalid_signature")
	}
	change, err := decodeEntitlementChange(body)
	if err != nil {
		return WebhookEvent{}, webhookFailure("invalid_event")
	}
	fingerprint := sha256.Sum256(body)
	duplicate, err := verifier.store.Remember(request.Context(), eventID, fingerprint[:])
	if err != nil {
		return WebhookEvent{}, wrapStoreError(err)
	}
	return WebhookEvent{ID: eventID, Timestamp: timestamp, Duplicate: duplicate, Change: change}, nil
}

// Close overwrites the in-memory signing secret.
func (verifier *WebhookVerifier) Close() {
	if verifier == nil {
		return
	}
	for index := range verifier.secret {
		verifier.secret[index] = 0
	}
}

// Remember inserts or compares one authenticated event identity.
func (store *MemoryEventStore) Remember(_ context.Context, eventID string, fingerprint []byte) (bool, error) {
	store.mu.Lock()
	defer store.mu.Unlock()
	if store.events == nil {
		store.events = make(map[string]string)
	}
	copied := string(fingerprint)
	existing, ok := store.events[eventID]
	if !ok {
		store.events[eventID] = copied
		return false, nil
	}
	if existing == copied {
		return true, nil
	}
	return false, webhookFailure("event_identity_conflict")
}

// validTimestamp parses a canonical Unix timestamp inside the configured replay window.
func (verifier *WebhookVerifier) validTimestamp(value string) (time.Time, string, bool) {
	if value == "" || strings.TrimSpace(value) != value {
		return time.Time{}, "", false
	}
	seconds, err := strconv.ParseInt(value, 10, 64)
	if err != nil || strconv.FormatInt(seconds, 10) != value {
		return time.Time{}, "", false
	}
	timestamp := time.Unix(seconds, 0).UTC()
	delta := verifier.clock().UTC().Sub(timestamp)
	if delta < -verifier.timestampTolerance || delta > verifier.timestampTolerance {
		return time.Time{}, "", false
	}
	return timestamp, value, true
}

// validSignature compares the provided HMAC-SHA-256 value in constant time.
func (verifier *WebhookVerifier) validSignature(eventID, timestamp string, body []byte, value string) bool {
	if strings.HasPrefix(value, "v1=") {
		return false
	}
	if !strings.HasPrefix(value, signaturePrefix) || len(value) != len(signaturePrefix)+(signatureBytes*2) {
		return false
	}
	provided, err := hex.DecodeString(strings.TrimPrefix(value, signaturePrefix))
	if err != nil || len(provided) != signatureBytes {
		return false
	}
	mac := hmac.New(sha256.New, verifier.secret)
	_, _ = io.WriteString(mac, signatureVersion)
	_, _ = io.WriteString(mac, signatureMACSeparator)
	_, _ = io.WriteString(mac, eventID)
	_, _ = io.WriteString(mac, signatureMACSeparator)
	_, _ = io.WriteString(mac, timestamp)
	_, _ = io.WriteString(mac, signatureMACSeparator)
	_, _ = mac.Write(body)
	return hmac.Equal(provided, mac.Sum(nil))
}

// decodeEntitlementChange validates the production entitlement.changed payload.
func decodeEntitlementChange(body []byte) (EntitlementChange, error) {
	var change EntitlementChange
	if err := json.Unmarshal(body, &change); err != nil {
		return EntitlementChange{}, err
	}
	if err := change.validate(); err != nil {
		return EntitlementChange{}, err
	}
	change.normalize()
	return change, nil
}

// isJSONContentType accepts JSON content types with optional parameters.
func isJSONContentType(value string) bool {
	mediaType, _, err := mime.ParseMediaType(value)
	return err == nil && mediaType == jsonContentType
}

// validWebhookIdentity accepts bounded printable identifiers without control bytes.
func validWebhookIdentity(value string) bool {
	if value == "" || len(value) > maximumWebhookIdentityLength || strings.TrimSpace(value) != value {
		return false
	}
	for _, character := range value {
		if character < 0x21 || character > 0x7e {
			return false
		}
	}
	return true
}

// webhookFailure constructs a WebhookError with the IAPStack retry-policy status.
func webhookFailure(code string) *WebhookError {
	return &WebhookError{Code: code, StatusCode: webhookStatus(code)}
}

// webhookStatus maps verifier failure codes onto IAPStack outbound delivery statuses.
func webhookStatus(code string) int {
	switch code {
	case "invalid_timestamp", "invalid_signature":
		return http.StatusUnauthorized
	case "method_not_allowed":
		return http.StatusMethodNotAllowed
	case "content_type_required":
		return http.StatusUnsupportedMediaType
	case "body_too_large":
		return http.StatusRequestEntityTooLarge
	case "event_identity_conflict":
		return http.StatusConflict
	case "receiver_unavailable":
		return http.StatusServiceUnavailable
	default:
		return http.StatusBadRequest
	}
}

// wrapStoreError preserves WebhookError values and maps other store failures to 503.
func wrapStoreError(err error) error {
	if err == nil {
		return nil
	}
	var webhookErr *WebhookError
	if errors.As(err, &webhookErr) {
		if webhookErr.StatusCode != 0 {
			return webhookErr
		}
		return &WebhookError{Code: webhookErr.Code, StatusCode: webhookStatus(webhookErr.Code), Cause: webhookErr.Cause}
	}
	return &WebhookError{Code: "receiver_unavailable", StatusCode: http.StatusServiceUnavailable, Cause: err}
}

// writeWebhookError writes the JSON error envelope hosts must return to IAPStack.
func writeWebhookError(writer http.ResponseWriter, err error) {
	webhookErr, ok := wrapStoreError(err).(*WebhookError)
	if !ok {
		webhookErr = webhookFailure("receiver_unavailable")
	}
	if webhookErr.Code == "method_not_allowed" {
		writer.Header().Set("Allow", http.MethodPost)
	}
	writer.Header().Set("Content-Type", jsonContentType)
	writer.WriteHeader(webhookErr.StatusCode)
	_ = json.NewEncoder(writer).Encode(map[string]string{"code": webhookErr.Code})
}

File: sdk/typescript/src/webhook.ts
import { createHash, createHmac, timingSafeEqual } from 'node:crypto';
import { WebhookError } from './errors';
import {
  EntitlementChange,
  optionalDate,
  requiredInt,
  requiredString,
} from './models';

const SIGNATURE_VERSION = 'v2';
const SIGNATURE_PREFIX = `${SIGNATURE_VERSION}=`;
const SIGNATURE_MAC_SEPARATOR = '\n';
const SIGNATURE_BYTES = 32;
const MINIMUM_WEBHOOK_SECRET_BYTES = 32;
const MAXIMUM_WEBHOOK_IDENTITY_LENGTH = 128;
const DEFAULT_WEBHOOK_BODY_LIMIT = 1 << 20;
const DEFAULT_WEBHOOK_TIMESTAMP_TOLERANCE_MS = 5 * 60 * 1000;
const JSON_CONTENT_TYPE = 'application/json';
const HEADER_EVENT_ID = 'IAPStack-Event-ID';
const HEADER_TIMESTAMP = 'IAPStack-Timestamp';
const HEADER_SIGNATURE = 'IAPStack-Signature';

/** Atomically deduplicates authenticated webhook event identities. */
export interface EventStore {
  /**
   * Records one authenticated event ID and body fingerprint.
   * Returns true when the same identity was stored before.
   * A conflicting fingerprint for an existing ID throws a WebhookError.
   */
  remember(eventId: string, fingerprint: Uint8Array): Promise<boolean> | boolean;
}

export interface WebhookConfig {
  /** Webhook signing secret; must contain at least 32 bytes. */
  secret: string | Uint8Array;
  /** Maximum accepted raw webhook body size. */
  bodyLimit?: number;
  /** Allowed clock skew around IAPStack-Timestamp, in milliseconds. */
  timestampToleranceMs?: number;
  /** Atomically deduplicates authenticated event IDs. */
  store: EventStore;
  /** Supplies the current time for replay-window checks. */
  clock?: () => Date;
}

export type WebhookHeaders = Headers | Record<string, string | string[] | undefined>;

/** Framework-agnostic authenticated webhook delivery. */
export interface WebhookRequest {
  method?: string;
  headers?: WebhookHeaders | null;
  body?: Uint8Array | ArrayBuffer | string | null;
}

/** One authenticated, optionally duplicate, entitlement change. */
export interface WebhookEvent {
  id: string;
  timestamp: Date;
  duplicate: boolean;
  change: EntitlementChange;
}

/** Process-local EventStore for tests and single-instance hosts. */
export class MemoryEventStore implements EventStore {
  private readonly events = new Map<string, Buffer>();

  remember(eventId: string, fingerprint: Uint8Array): boolean {
    const next = Buffer.from(fingerprint);
    const existing = this.events.get(eventId);
    if (!existing) {
      this.events.set(eventId, next);
      return false;
    }
    if (existing.length === next.length && timingSafeEqual(existing, next)) {
      return true;
    }
    throw new WebhookError('event_identity_conflict');
  }
}

/** Authenticates v2 IAPStack deliveries and deduplicates event IDs. */
export class WebhookVerifier {
  private secret: Buffer;
  private readonly bodyLimit: number;
  private readonly timestampToleranceMs: number;
  private readonly store: EventStore;
  private readonly clock: () => Date;

  constructor(config: WebhookConfig) {
    const secret = toSecretBuffer(config.secret);
    if (secret.length < MINIMUM_WEBHOOK_SECRET_BYTES) {
      throw new TypeError('webhook signing secret must contain at least 32 bytes');
    }
    if (!config.store) {
      throw new TypeError('webhook event store is required');
    }
    const bodyLimit = config.bodyLimit ?? DEFAULT_WEBHOOK_BODY_LIMIT;
    if (bodyLimit <= 0) {
      throw new TypeError('webhook body limit must be positive');
    }
    const timestampToleranceMs =
      config.timestampToleranceMs ?? DEFAULT_WEBHOOK_TIMESTAMP_TOLERANCE_MS;
    if (timestampToleranceMs <= 0) {
      throw new TypeError('webhook timestamp tolerance must be positive');
    }
    this.secret = Buffer.from(secret);
    this.bodyLimit = bodyLimit;
    this.timestampToleranceMs = timestampToleranceMs;
    this.store = config.store;
    this.clock = config.clock ?? (() => new Date());
  }

  /**
   * Authenticates one HTTP delivery, parses the payload, and deduplicates it.
   * Pass the exact raw body; do not re-serialize parsed JSON.
   */
  async verifyRequest(request: WebhookRequest | Request | null | undefined): Promise<WebhookEvent> {
    if (!request) {
      throw new WebhookError('invalid_body');
    }
    const method = requestMethod(request);
    const headers = requestHeaders(request);
    if (method !== 'POST') {
      throw new WebhookError('method_not_allowed');
    }
    if (!isJSONContentType(headerValue(headers, 'Content-Type'))) {
      throw new WebhookError('content_type_required');
    }
    const eventId = headerValue(headers, HEADER_EVENT_ID);
    if (!validWebhookIdentity(eventId)) {
      throw new WebhookError('invalid_event_id');
    }
    const timestampHeader = headerValue(headers, HEADER_TIMESTAMP);
    const timestamp = this.validTimestamp(timestampHeader);
    if (!timestamp) {
      throw new WebhookError('invalid_timestamp');
    }
    const body = await readRequestBody(request, this.bodyLimit);
    if (
      !this.validSignature(eventId, timestampHeader, body, headerValue(headers, HEADER_SIGNATURE))
    ) {
      throw new WebhookError('invalid_signature');
    }
    const change = decodeEntitlementChange(body);
    const fingerprint = createHash('sha256').update(body).digest();
    const duplicate = await this.store.remember(eventId, fingerprint);
    return {
      id: eventId,
      timestamp,
      duplicate,
      change,
    };
  }

  /** Overwrites the in-memory signing secret. */
  close(): void {
    this.secret.fill(0);
  }

  private validTimestamp(value: string): Date | null {
    if (!value || value.trim() !== value) {
      return null;
    }
    if (!/^-?\d+$/u.test(value)) {
      return null;
    }
    let seconds: bigint;
    try {
      seconds = BigInt(value);
    } catch {
      return null;
    }
    if (seconds.toString() !== value) {
      return null;
    }
    if (seconds > BigInt(Number.MAX_SAFE_INTEGER) || seconds < BigInt(Number.MIN_SAFE_INTEGER)) {
      return null;
    }
    const timestamp = new Date(Number(seconds) * 1000);
    const delta = this.clock().getTime() - timestamp.getTime();
    if (delta < -this.timestampToleranceMs || delta > this.timestampToleranceMs) {
      return null;
    }
    return timestamp;
  }

  private validSignature(eventId: string, timestamp: string, body: Buffer, value: string): boolean {
    if (value.startsWith('v1=')) {
      return false;
    }
    if (!value.startsWith(SIGNATURE_PREFIX) || value.length !== SIGNATURE_PREFIX.length + SIGNATURE_BYTES * 2) {
      return false;
    }
    const providedHex = value.slice(SIGNATURE_PREFIX.length);
    if (!/^[0-9a-fA-F]+$/u.test(providedHex)) {
      return false;
    }
    const provided = Buffer.from(providedHex, 'hex');
    if (provided.length !== SIGNATURE_BYTES) {
      return false;
    }
    const mac = createHmac('sha256', this.secret);
    mac.update(SIGNATURE_VERSION);
    mac.update(SIGNATURE_MAC_SEPARATOR);
    mac.update(eventId);
    mac.update(SIGNATURE_MAC_SEPARATOR);
    mac.update(timestamp);
    mac.update(SIGNATURE_MAC_SEPARATOR);
    mac.update(body);
    const expected = mac.digest();
    return expected.length === provided.length && timingSafeEqual(expected, provided);
  }
}

function requestMethod(request: WebhookRequest | Request): string {
  return isFetchRequest(request) ? request.method : (request.method ?? '');
}

function requestHeaders(request: WebhookRequest | Request): WebhookHeaders {
  return isFetchRequest(request) ? request.headers : (request.headers ?? {});
}

async function readRequestBody(request: WebhookRequest | Request, bodyLimit: number): Promise<Buffer> {
  if (isFetchRequest(request)) {
    return readLimitedStream(request, bodyLimit);
  }
  if (request.body === null || request.body === undefined) {
    throw new WebhookError('invalid_body');
  }
  const body = toBodyBuffer(request.body);
  if (body.length > bodyLimit) {
    throw new WebhookError('body_too_large');
  }
  return body;
}

async function readLimitedStream(request: Request, bodyLimit: number): Promise<Buffer> {
  const body = request.body;
  if (!body || typeof body.getReader !== 'function') {
    const buffer = Buffer.from(await request.arrayBuffer());
    if (buffer.length > bodyLimit) {
      throw new WebhookError('body_too_large');
    }
    return buffer;
  }
  const reader = body.getReader();
  const chunks: Uint8Array[] = [];
  let length = 0;
  for (;;) {
    const { done, value } = await reader.read();
    if (done) {
      break;
    }
    if (!value) {
      continue;
    }
    length += value.byteLength;
    if (length > bodyLimit) {
      try {
        await reader.cancel();
      } catch {
        // Size cap already exceeded; cancel is best-effort.
      }
      throw new WebhookError('body_too_large');
    }
    chunks.push(value);
  }
  return Buffer.concat(chunks, length);
}

function isFetchRequest(request: WebhookRequest | Request): request is Request {
  return typeof Request !== 'undefined' && request instanceof Request;
}

function toBodyBuffer(body: Uint8Array | ArrayBuffer | string): Buffer {
  if (typeof body === 'string') {
    return Buffer.from(body);
  }
  if (body instanceof ArrayBuffer) {
    return Buffer.from(body);
  }
  return Buffer.from(body);
}

function toSecretBuffer(secret: string | Uint8Array): Buffer {
  return typeof secret === 'string' ? Buffer.from(secret) : Buffer.from(secret);
}

function headerValue(headers: WebhookHeaders | null | undefined, name: string): string {
  if (!headers) {
    return '';
  }
  if (typeof Headers !== 'undefined' && headers instanceof Headers) {
    return headers.get(name) ?? '';
  }
  const record = headers as Record<string, string | string[] | undefined>;
  const direct = record[name] ?? record[name.toLowerCase()];
  if (Array.isArray(direct)) {
    return direct[0] ?? '';
  }
  if (typeof direct === 'string') {
    return direct;
  }
  const wanted = name.toLowerCase();
  for (const [key, value] of Object.entries(record)) {
    if (key.toLowerCase() === wanted) {
      if (Array.isArray(value)) {
        return value[0] ?? '';
      }
      return value ?? '';
    }
  }
  return '';
}

function isJSONContentType(value: string): boolean {
  const mediaType = parseMediaType(value);
  return mediaType === JSON_CONTENT_TYPE;
}

function parseMediaType(value: string): string {
  const [raw] = value.split(';', 1);
  return (raw ?? '').trim().toLowerCase();
}

function validWebhookIdentity(value: string): boolean {
  if (!value || value.length > MAXIMUM_WEBHOOK_IDENTITY_LENGTH || value.trim() !== value) {
    return false;
  }
  for (let index = 0; index < value.length; index += 1) {
    const code = value.charCodeAt(index);
    if (code < 0x21 || code > 0x7e) {
      return false;
    }
  }
  return true;
}

function decodeEntitlementChange(body: Buffer): EntitlementChange {
  let payload: unknown;
  try {
    payload = JSON.parse(body.toString('utf8'));
  } catch {
    throw new WebhookError('invalid_event');
  }
  if (!payload || typeof payload !== 'object' || Array.isArray(payload)) {
    throw new WebhookError('invalid_event');
  }
  const json = payload as Record<string, unknown>;
  try {
    const schemaVersion = requiredInt(json, 'schema_version');
    const version = requiredInt(json, 'version');
    const change = new EntitlementChange({
      schemaVersion,
      projectId: requiredString(json, 'project_id'),
      applicationId: requiredString(json, 'application_id'),
      customerId: requiredString(json, 'customer_id'),
      entitlementId: requiredString(json, 'entitlement_id'),
      entitlementKey: requiredString(json, 'entitlement_key'),
      access: requiredString(json, 'access'),
      accessReason: requiredString(json, 'access_reason'),
      sourceObservationId: requiredString(json, 'source_observation_id'),
      sourceApplicationId: requiredString(json, 'source_application_id'),
      sourceProductId: requiredString(json, 'source_product_id'),
      effectiveStartsAt: optionalDate(json, 'effective_starts_at'),
      effectiveEndsAt: optionalDate(json, 'effective_ends_at'),
      version,
    });
    if (change.schemaVersion <= 0 || change.version < 1) {
      throw new Error('webhook event metadata is invalid');
    }
    return change;
  } catch {
    throw new WebhookError('invalid_event');
  }
}
